### PR TITLE
RCHAIN-3643 introduce key in pyrchain

### PR DESCRIPTION
**Pull request by zsluedem**
## Overview
introduce key in pyrchain

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3643

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by zsluedem** [comment]
bors try

File: integration-tests/test/test_wallets.py [hunk 1]
```diff
@@ -86,41 +86,41 @@ def get_rev_address_by_public_key(context: TestingContext, node: Node, public_ke
 def test_alice_pay_bob(started_standalone_bootstrap_node: Node, command_line_options: CommandLineOptions, docker_client: DockerClient, random_generator: Random) -> None:
     with testing_context(command_line_options, random_generator, docker_client) as context:
         transfer_amount = 100
-        alice_rev_address = get_rev_address_by_public_key(context, started_standalone_bootstrap_node, ALICE_KEY.public_key, ALICE_KEY.private_key)
-        bob_rev_address = get_rev_address_by_public_key(context, started_standalone_bootstrap_node, BOB_KEY.public_key, BOB_KEY.private_key)
-        create_genesis_vault(context, started_standalone_bootstrap_node, alice_rev_address, ALICE_GENESIS_VAULT_AMOUNT, ALICE_KEY.private_key)
-        alice_balance = get_vault_balance(context, started_standalone_bootstrap_node, alice_rev_address, ALICE_KEY.private_key)
-        bob_balance = get_vault_balance(context, started_standalone_bootstrap_node, bob_rev_address, BOB_KEY.private_key)
+        alice_rev_address = get_rev_address_by_public_key(context, started_standalone_bootstrap_node, ALICE_KEY.get_public_key().to_hex(), ALICE_KEY.to_hex())
```
Comment (adaszko) [inline integration-tests/test/test_wallets.py]: The `.to_hex()` method should be called right before interpolating the key into a string, not before passing it as a parameter.  Functions like `get_rev_address_by_public_key()` should accept the `PrivateKey` type, not `str`.

**Review by adaszko — CHANGES_REQUESTED** [review]

File: integration-tests/Pipfile [hunk 2]
```diff
@@ -16,7 +16,7 @@ ecdsa = "*"
 cryptography = "*"
 eth-hash = "*"
 pycryptodome = "*"
-pyrchain = "*"
+pyrchain = {git = "https://github.com/rchain/pyrchain.git"}
```
Comment (adaszko) [inline integration-tests/Pipfile:19]: Good idea 👍

**Comment by adaszko** [comment]
Is the public key derived from the private key now?

**Comment by woky** [comment]
> Is the public key derived from the private key now?

Correct. (`get_public_key()`)

**Comment by zsluedem** [comment]
bors try

File: integration-tests/test/test_dag_correctness.py [hunk 3]
```diff
@@ -34,22 +35,22 @@ def test_fault_tolerance(command_line_options: CommandLineOptions, docker_client
...
-                    with bootstrap_connected_peer(context=context, bootstrap=bootstrap_node, name='bonded-validator-3', keypair=BONDED_VALIDATOR_KEY_3) as validator3:
+            with bootstrap_connected_peer(context=context, bootstrap=bootstrap_node, name='bonded-validator-1', private_key=BONDED_VALIDATOR_KEY_1) as validator1:
+                with bootstrap_connected_peer(context=context, bootstrap=bootstrap_node, name='bonded-validator-2', private_key=BONDED_VALIDATOR_KEY_2) as validator2:
+                    with bootstrap_connected_peer(context=context, bootstrap=bootstrap_node, name='bonded-validator-3', private_key=BONDED_VALIDATOR_KEY_3) as validator3:
                         wait_for_peers_count_at_least(context, validator1, 3)
                         wait_for_peers_count_at_least(context, validator2, 3)
                         wait_for_peers_count_at_least(context, validator3, 3)
 
-                        deploy1 = DeployThread("validator1", validator1, contract_path, 10, BONDED_VALIDATOR_KEY_1.private_key)
+                        deploy1 = DeployThread("validator1", validator1, contract_path, 10, BONDED_VALIDATOR_KEY_1.to_hex())
```
Comment (adaszko) [inline integration-tests/test/test_dag_correctness.py]: Same here

File: integration-tests/test/test_propose.py [hunk 4]
```diff
@@ -1,21 +1,21 @@
...
-DEPLOY_KEY = KeyPair(private_key='632a21e0176c4daed1ca78f08f98885f61d2050e0391e31eae59ff1a35ccca7f', public_key='040d09c2c290d458d666df9be22fe77cc71711bc052656bc089662c803ad61568a647585acda4974ea63dd7f820d1f349a498684e2527941140c5e0386441a2177')
+DEPLOY_KEY = PrivateKey.from_hex("632a21e0176c4daed1ca78f08f98885f61d2050e0391e31eae59ff1a35ccca7f")
 
 
 def test_propose(started_standalone_bootstrap_node: Node, random_generator: Random) -> None:
     relative_paths = started_standalone_bootstrap_node.shell_out('sh', '-c', 'ls /opt/docker/examples/*.rho').splitlines()
     relative_path = random_generator.choice(relative_paths)
     full_path = os.path.join('/opt/docker/examples', relative_path)
-    started_standalone_bootstrap_node.deploy(full_path, DEPLOY_KEY.private_key)
+    started_standalone_bootstrap_node.deploy(full_path, DEPLOY_KEY.to_hex())
```
Comment (adaszko) [inline integration-tests/test/test_propose.py]: and here

File: integration-tests/test/test_slash.py [hunk 5]
```diff
@@ -36,14 +38,14 @@ def hex_block_hash(block_hash: bytes) -> bytes:
...
         docker_network_with_started_bootstrap(context=context) as bootstrap_node, \
-        bootstrap_connected_peer(context=context, bootstrap=bootstrap_node, name='bonded-validator-1', keypair=BONDED_VALIDATOR_KEY_1) as validator1, \
-        bootstrap_connected_peer(context=context, bootstrap=bootstrap_node, name='bonded-validator-2', keypair=BONDED_VALIDATOR_KEY_2) as validator2, \
+        bootstrap_connected_peer(context=context, bootstrap=bootstrap_node, name='bonded-validator-1', private_key=BONDED_VALIDATOR_KEY_1) as validator1, \
+        bootstrap_connected_peer(context=context, bootstrap=bootstrap_node, name='bonded-validator-2', private_key=BONDED_VALIDATOR_KEY_2) as validator2, \
         node_protocol_client(bootstrap_node.network, docker_client) as client:
             contract = '/opt/docker/examples/tut-hello.rho'
 
-            validator1.deploy(contract, BONDED_VALIDATOR_KEY_1.private_key)
+            validator1.deploy(contract, BONDED_VALIDATOR_KEY_1.to_hex())
```
Comment (adaszko) [inline integration-tests/test/test_slash.py]: and here

**Review by adaszko — CHANGES_REQUESTED** [review]

**Comment by KentShikama** [comment]
This should make https://github.com/rchain/rchain/pull/2476 pass right?

**Comment by zsluedem** [comment]
bors try

**Comment by zsluedem** [comment]
> This should make #2476 pass right?

I don't think so. All the tests related to `deploy and propose` fail in that pr.

**Comment by KentShikama** [comment]
@zsluedem Nvm, I confused this PR with a ticket that would sign deploys in integration tests.

`INFO     root:rnode.py:287 rchain-gcjpe.bootstrap: Failure! Invalid block: DeployNotSigned`

**Comment by KentShikama** [comment]
@zsluedem Is signing deploys used in integration tests somewhere on your TODO list though?

**Comment by zsluedem** [comment]
> @zsluedem Is signing deploys used in integration tests somewhere on your TODO list though?

I think https://github.com/rchain/pyrchain/blob/master/rchain/util.py#L18-L38 is something you said.

**Comment by woky** [comment]
> > @zsluedem Is signing deploys used in integration tests somewhere on your TODO list though?
> 
> I think https://github.com/rchain/pyrchain/blob/master/rchain/util.py#L18-L38 is something you said.

Yeah, if int. tests switch to Pyrchain and use `rclient.deploy(...)`, deploys should be signed.

**Comment by zsluedem** [comment]
@adaszko updated~!

File: integration-tests/test/test_dag_correctness.py (hunk 3 above)
Comment (adaszko) [reply integration-tests/test/test_dag_correctness.py]: Can you make `DeployThread.__init__()` take `PrivateKey` type in lieu of `str`?

**Review by adaszko — CHANGES_REQUESTED** [review]

**Review by adaszko — APPROVED** [review]

**Review by woky — APPROVED** [review]

**Pull request merged**